### DISCUSSION TITLE
How to keep string literals in the instructions when using getInstructions() in script?

### DISCUSSION THREAD

**Discussion by fzuo** [NationalSecurityAgency/ghidra#5648] [Q&A] [answered]
Take this 'hello world' program as an example, here is from the dissembler window of the ghidra
 
        00101149 f3 0f 1e fa     ENDBR64
        0010114d 55                  PUSH       RBP
        0010114e 48 89 e5        MOV        RBP,RSP
        00101151 48 8d 05        LEA        RAX,[s_Hello,_World!_00102004]                   = "Hello, World!"
                         ac 0e 00 00
        00101158 48 89 c7        MOV        RDI=>s_Hello,_World!_00102004,RAX                = "Hello, World!"
        0010115b e8 f0 fe          CALL       <EXTERNAL>::puts                                 int puts(char * __s)
                         ff ff
        00101160 b8 00 00        MOV        EAX,0x0
                         00 00
        00101165 5d                  POP        RBP
        00101166 c3                  RET

But, if I use getInstructions(block, True), below is the output:

00101149 ENDBR64
0010114d PUSH RBP
0010114e MOV RBP,RSP
00101151 LEA RAX,[0x102004]
00101158 MOV RDI,RAX
0010115b CALL 0x00101050
00101160 MOV EAX,0x0
00101165 POP RBP
00101166 RET

The string literal "Hello, World!" is replaced with an address [0x102004]. 

How can I keep the string literal in the output in script?

**Comment by ghidracadabra** [comment]
You can try passing each instruction to `CodeUnitFormat.DEFAULT.getRepresentationString(CodeUnit)`

If the `DEFAULT` format isn't sufficient, there are constructors in the `CodeUnitFormat` class that allow you to set various options.

**Reply by fzuo** [reply]
Thanks for your help. I just have a follow-up question. By using the script below:

    ins_iter = listing.getInstructions(block, True)
 
    while ins_iter.hasNext():
        ins = ins_iter.next()
        cu = listing.getCodeUnitAt(ins.getAddress())
        deol = DisplayableEol(cu, True, True, True, True, 5, True, True)
        if deol.hasAutomatic():
            ac = deol.getAutomaticComment()
            print(ac[0])
        formatted_ins = CodeUnitFormat.DEFAULT.getRepresentationString(ins)
        print(formatted_ins)

I can get the output as below:
ENDBR64
PUSH RBP
MOV RBP,RSP
**= "Hello, World!"**
LEA RAX,[**s_Hello,_World!_00002004**]
= "Hello, World!"
MOV RDI,RAX
int puts(char * __s)
CALL puts
MOV EAX,0x0
POP RBP
RET

[**s_Hello,_World!_00002004**] is a formatted instruction, while **= "Hello, World!"** is an automatic comment. Is there anyway that can allow me to replace [**s_Hello,_World!_00002004**]  with "Hello, World!" directly. If ghidra can generate a comment, I suppose there should be a possible solution. I cannot use the comment directly, because for MOV RDI,RAX, it also has a similar comment **= "Hello, World!"**.

Thanks in advance!

**Reply by ghidracadabra** [reply]
I don't think there is a way to do this; you'd have to do the string manipulation yourself.

**Comment by fzuo** [comment]
How can I know an operand such as [0x102004] actually refers to a string?

**Comment by ghidracadabra** [comment]
Here is a snippet of a ghidra script that you should be able to adapt.


```
Instruction inst = getInstructionContaining(currentAddress);
if (inst == null) {
    return;
}
Reference[] refs = inst.getReferencesFrom();
for (Reference ref : refs) {
    if (ref.getOperandIndex() == ReferenceManager.MNEMONIC) {
        continue;
    }
    if (!ref.getReferenceType().equals(RefType.DATA)) {
        continue;
    }
    Address target = ref.getToAddress();
    Data d = getDataAt(target);
    if (d != null && d.getDataType() instanceof StringDataType) {
        printf("Operand %d references string %s\n", ref.getOperandIndex(), d.getDefaultValueRepresentation());
    }
}
```

**Reply by dev747368** [reply, marked as answer]
```StringDataInstance.isString( d )``` is a slightly better test as it also handles char[] arrays as well as stuff in the StringDataType flavor.  (however, the test would be better if it was ```instanceof AbstractStringDataType```.

Use ```StringDataInstance.getStringDataInstance( d )```  to be able to access the data there as a string.

**Reply by fzuo** [reply]
Thanks for your help!
